packages/xstate-react/src/useInterpret.ts

```diff
@@ -0,0 +1,110 @@
+import { useState, useEffect } from 'react';
+import useIsomorphicLayoutEffect from 'use-isomorphic-layout-effect';
+import {
+  interpret,
+  EventObject,
+  StateMachine,
+  State,
+  Interpreter,
+  InterpreterOptions,
+  MachineOptions,
+  Typestate,
+  Observer
+} from 'xstate';
+import { toObserver } from 'xstate/lib/utils';
+import { MaybeLazy } from './types';
+import useConstant from './useConstant';
+import { UseMachineOptions } from './useMachine';
+import { useReactEffectActions } from './useReactEffectActions';
+
+export function useInterpret<
+  TContext,
+  TEvent extends EventObject,
+  TTypestate extends Typestate<TContext> = { value: any; context: TContext }
+>(
+  getMachine: MaybeLazy<StateMachine<TContext, any, TEvent, TTypestate>>,
+  options: Partial<InterpreterOptions> &
+    Partial<UseMachineOptions<TContext, TEvent>> &
+    Partial<MachineOptions<TContext, TEvent>> = {},
+  observerOrListener?:
+    | Observer<State<TContext, TEvent, any, TTypestate>>
+    | ((value: State<TContext, TEvent, any, TTypestate>) => void)
+): Interpreter<TContext, any, TEvent, TTypestate> {
+  const machine = useConstant(() => {
+    return typeof getMachine === 'function' ? getMachine() : getMachine;
+  });
+
+  if (
+    process.env.NODE_ENV !== 'production' &&
+    typeof getMachine !== 'function'
+  ) {
+    const [initialMachine] = useState(machine);
+
+    if (machine !== initialMachine) {
+      console.warn(
+        'Machine given to `useMachine` has changed between renders. This is not supported and might lead to unexpected results.\n' +
+          'Please make sure that you pass the same Machine as argument each time.'
+      );
+    }
+  }
+
+  const {
+    context,
+    guards,
+    actions,
+    activities,
+    services,
+    delays,
+    state: rehydratedState,
+    ...interpreterOptions
+  } = options;
+
+  const service = useConstant(() => {
+    const machineConfig = {
+      context,
+      guards,
+      actions,
+      activities,
+      services,
+      delays
+    };
+    const machineWithConfig = machine.withConfig(machineConfig, {
+      ...machine.context,
+      ...context
+    } as TContext);
+
+    return interpret(machineWithConfig, {
+      deferEvents: true,
+      ...interpreterOptions
+    });
+  });
+
+  useIsomorphicLayoutEffect(() => {
+    if (observerOrListener) {
+      service.subscribe(toObserver(observerOrListener));
+    }
+    service.start(
+      rehydratedState ? (State.create(rehydratedState) as any) : undefined
+    );
+
+    return () => {
+      service.stop();
+    };
+  }, []);
+
+  // Make sure machine options are kept updated when they change.
+  // This mutation assignment is safe because the service instance is only used
+  // in one place -- this hook's caller.
+  useEffect(() => {
+    Object.assign(service.machine.options, {
+      guards,
```

Comment:
Looking at the implementation of XState, I don't think updating `guards` will actually work - guards are currently destructured in `formatTransition`:
https://github.com/davidkpiano/xstate/blob/a9ab426b53e416c776af514b7b3a4f6ef3f0d855/packages/core/src/StateNode.ts#L1826
and "embedded" in the transition definition here:
https://github.com/davidkpiano/xstate/blob/a9ab426b53e416c776af514b7b3a4f6ef3f0d855/packages/core/src/StateNode.ts#L1833
https://github.com/davidkpiano/xstate/blob/a9ab426b53e416c776af514b7b3a4f6ef3f0d855/packages/core/src/utils.ts#L495-L501

and the unpackaged `.predicate` is used here in the `evaluateGuard`:
https://github.com/davidkpiano/xstate/blob/a9ab426b53e416c776af514b7b3a4f6ef3f0d855/packages/core/src/utils.ts#L659-L665